### PR TITLE
OcclusionQueryNode: fix use case of user defined query geometry

### PR DESCRIPTION
**Pull request by dan-t**
The user defined query geometry handling has been broken in several ways.

The previous way of defining a query geometry was using the non const
`getQueryGeometry` method and overriding its members. But then
`OcclusionQueryNode` wasn't aware of the geometry change and couldn't
internally handle it correctly.

The `computeBound` method never considered a user defined query geometry and
always just overrode the vertices of the geometry.

The member `_validQueryGeometry` wasn't correctly set.

This change should fix all this issues by introducing a small backward
compatibility break. The non const `getQueryGeometry` method is removed
forcing the user to use the `setQueryGeometry` method. But then `OcclusionQueryNode`
is aware of the user defined query geometry and can handle it correctly.

### CODE REVIEW

**Comment by openscenegraph** [comment]
@mp3butcher Julian, could you review this PR and see if it's addresses the issues you were up against?

**Comment by mp3butcher** [comment]
a pr against 3.6 would be better

**Comment by mp3butcher** [comment]
...master lacks a commit i've done to take into acount _enable

**Comment by mp3butcher** [comment]
@robertosfield 
https://github.com/openscenegraph/OpenSceneGraph/commit/cfd190a9faf886f3bf9b5e2c44ef0ceef2631202
and
https://github.com/openscenegraph/OpenSceneGraph/commit/5311d6fb6b1a1840322e12a52d58792193f81e61
may be merge in master
before this one

**Comment by mp3butcher** [comment]
@dan-t removing not const getQuerYGeometry breaks backward client usage

File: include/osg/OcclusionQueryNode [hunk 1]
```diff
@@ -158,8 +158,11 @@ public:
     osg::StateSet* getQueryStateSet();
     const osg::StateSet* getQueryStateSet() const;
 
-    // Get the QueryGeometry object used for occlusion query. Returns 0 if no QueryGeometry is created.
-    osg::QueryGeometry* getQueryGeometry();
```
[the remark this answers is not in the document]
Comment (dan-t) [reply include/osg/OcclusionQueryNode]: @robertosfield Can you please decide if breaking backward compatibility is acceptable if it fixes a broken version.

**Comment by openscenegraph** [comment]
Hi Dan,

I'm happy to see a breaking change if it addresses problem usage cases.  I
expect only a small % of users will be using occlussion query and those who
do will sure want to have their usage working correctly.

Cheers,
Robert.

File: include/osg/OcclusionQueryNode (hunk 1 above)
[the remark this answers is not in the document]
Comment (mp3butcher) [reply include/osg/OcclusionQueryNode]: @dan-t my pr doesn't break backward compatbility and set validgeometry to true...take a look i have mod it

[the remark this answers is not in the document]
Comment (mp3butcher) [reply include/osg/OcclusionQueryNode]: as i already mention, the introduction of validgeometry  breaks backward compatibility in all cases...

[the remark this answers is not in the document]
Comment (mp3butcher) [reply include/osg/OcclusionQueryNode]: something like 
```
 if ( !_validQueryGeometry )
    {
        if(qg->getBoundingBox().valid())
            _validQueryGeometry = true;
        else
        {
            // There're cases that the occlusion test result has been retrieved
            // after the query geometry has been changed, it's the result of the
            // geometry before the change.
            qg->reset();
    
            // The box of the query geometry is invalid, return false to not traverse
            // the subgraphs.
            _passed = false;
            return _passed;
        }
    }
```

**Comment by mp3butcher** [comment]
https://github.com/openscenegraph/OpenSceneGraph/pull/811/commits/1a9fcfc126332f03e7c3c4ec3f2b5788807732fc
solve the backward compat issue

**Comment by dan-t** [comment]
> 1a9fcfc
> solve the backward compat issue
> 
I‘m sorry, but that‘s just a hack.

When you keep the non const ˋgetQueryGeometryˋ, you can‘t fix ˋcomputeBoundˋ , it just overrides the vertices of the user define query geometry. It even isn‘t thread safe to change the query geometry, that‘s why ˋsetQueryGeometryInternalˋ locks the
ˋ_computeBoundMutexˋ.

**Comment by mp3butcher** [comment]
oups u're right (validgeometry is not muted properly)...
I'm pretty ok with your changes but as long as your pr is not against 3.6 branch or(@robertosfield  don't merge https://github.com/openscenegraph/OpenSceneGraph/pull/772 with master) i won't be happy with it

**Comment by dan-t** [comment]
> I'm pretty ok with your changes but as long as your pr is not against 3.6 branch or(@robertosfield don't merge #772 with master) i won't be happy with it
> 
For your testing purposes you can just ˋgit cherry-pickˋ the commits
onto your local branch.

**Comment by mp3butcher** [comment]
as i already said cherry-pick this commit into 3.6 branch involve conflicts so for proper reviewing something should be done by you (pr against 3.6) or Robert (merging https://github.com/openscenegraph/OpenSceneGraph/pull/772)

**Comment by openscenegraph** [comment]
I've merged the || !_enabled change in 3.6 into master:

void OcclusionQueryNode::traverseQuery( const Camera* camera, NodeVisitor& nv )
{
    if (!_validQueryGeometry || ! _enabled)
        return;

I don't know how it ended up diverged from 3.6, one missing commit me thinks.  I will now merge this PR as it look generally OK, but I'm @dan-t lots here as I don't have too much knowledge about this particular class and it's usage.

**Comment by openscenegraph** [comment]
Ack, now got a compile error with master, the above tweak of mine wasn't appropriate... now looking into how to resolve it.

**Comment by openscenegraph** [comment]
Hang my head in shame, I broke the build be inappropriate trying to match the 3.6 and master on this PR without spotting all the changes.  Now fixed:

https://github.com/openscenegraph/OpenSceneGraph/commit/3a17fd3e9c6be5272b4f89b6849a8ca9251f5215

**Comment by mp3butcher** [comment]
@openscenegraph please merge with 3.6 branch

**Pull request merged**